### ISSUE TITLE
Stalls in the middle of model generation and will run forever

### ISSUE THREAD

**Issue by mazeau** [cfgoldsmith/RMG-Py#69]
Sometimes the model generation will stall after
```
Summary of Model Enlargement
---------------------------------
Added 1 new core species
    [CH]=C=C=C(C)C=C=C=C=C(28628)
Created 0 new edge species
Moved 1 reactions from edge to core
    C#CC#CC(23824) + C#CC#C[CH2](26090) <=> [CH]=C=C=C(C)C=C=C=C=C(28628)
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 200 species and 2234 reactions
    The model edge has 43657 species and 108156 reactions
```
in the RMG.log.  This specific RMG.log was last written at 9:41 AM (as well as the chemkin folders and the solver folder) and seems to just stop writing to anywhere and doing anything until it times out, then the error.log file is written (1:44 PM). 

The next step should be to write `Adding reaction library Surface/CPOX_Pt/Deutschmann2006 to output file...`, but for some reason can't?  Searching for the new species 28628 in the output.html yields nothing.

This only happens for some models with -1.5 to -3.5 carbon binding energies.

**Comment by mazeau** [comment]
some profiles of models that were terminated before they could finish
dumping the ones that at first glance aren't duplicates here before I overwrite them
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765129/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765133/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765135/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765139/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765141/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765143/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765146/RMG.profile.dot.pdf)
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765149/RMG.profile.dot.pdf)


one that was able to finish:
[RMG.profile.dot.pdf](https://github.com/cfgoldsmith/RMG-Py/files/3765158/RMG.profile.dot.pdf)

**Comment by mliu49** [comment]
It would be helpful to know the kinds of molecules being simulated and how far each of the jobs got for the different profiles.

That aside, I just looked at the last one, which had 87% time in `find_degenerate_reactions` and 45% time in `ensure_species`. I would definitely say that this is not normal, and my first guess is that there is a species with a ton of resonance structures - `ensure_species` is the first time that resonance structures are generated for new species, and `find_degenerate_reactions` does isomorphism which would be slow if there are species with tons of resonance structures.

The species added to core mentioned in the first comment `[CH]=C=C=C(C)C=C=C=C=C` only has 2 resonance structures though. Does the model also contain heteroatoms? It is possible that that species combined with a heteroatom with lone pairs could result in a ton of resonance structures.

**Comment by mazeau** [comment]
No, there is only C and H in the system.  This is oligomerization/dimerization of ethylene with n-heptane as a solvent.  I also ran into this in some CPOX models

**Comment by rwest** [comment]
The opening comment of this issue may not be relevant to these profiles - it is 7 months old, and I forget what we were even simulating then. Emilie, could you share the input file and perhaps the log for one of these? Might be informative to see what was added to the core just before the end

**Comment by rwest** [comment]
Worth saying: this is with heterogeneous surfaces. Not sure our resonance stuff has been well tested there

**Comment by mazeau** [comment]
[input.py.txt](https://github.com/cfgoldsmith/RMG-Py/files/3765272/input.py.txt)
[RMG.log](https://github.com/cfgoldsmith/RMG-Py/files/3765273/RMG.log)

**Comment by rwest** [comment]
Lol. Look at those edge species!

![image](https://user-images.githubusercontent.com/93807/67446735-8fc3d980-f5df-11e9-9d63-7f6e7a458e1d.png)
![image](https://user-images.githubusercontent.com/93807/67446751-9e11f580-f5df-11e9-8fb4-8a75afdec6e5.png)
![image](https://user-images.githubusercontent.com/93807/67446866-ff39c900-f5df-11e9-9d22-81af058b111e.png)

I know the point is to model oligemerization (and hence some polymerization, as an unwanted but important pathway) but perhaps we could put some [constraints](http://reactionmechanismgenerator.github.io/RMG-Py/users/rmg/input.html#species-constraints) on the size of a molecule, and also the number of radicals it contains,

**Comment by mliu49** [comment]
It's possible that this isn't due to a couple molecules with tons of resonance structures, but the accumulation of tons of molecules with many resonance structures. It's hard to say for sure though.

I think limiting the number of radicals would likely help.

**Comment by rwest** [comment]
Just following up here, in case we refer to this thread again in future.

Analyzing the dictionary of species from the edge of one of Emily's models that ran for a long time with a huge majority of time spent generating and comparing resonance structures...

Number of resonance structures :   Number of species of this type
                          1   :   457
                          2   :    82
                          3   :   546
                          4   :    47
                          5   :     7
                          6   :   352
                          7   :     1
                          8   :     9
                          9   :   270
                         10   :     2
                         12   :    52
                         15   :     2
                         16   :     1
                         18   :   204
                         24   :     5
                         27   :    79
                         36   :    50
                         54   :    25
                         81   :    15

Here are some of the ones with 54 resonance structures:
![image](https://user-images.githubusercontent.com/93807/67952297-7b757300-fbc3-11e9-9ecb-cc8da873847e.png)
![image](https://user-images.githubusercontent.com/93807/67952309-816b5400-fbc3-11e9-8d94-1b90db8677c5.png)
![image](https://user-images.githubusercontent.com/93807/67952179-449f5d00-fbc3-11e9-8dfa-faa1bae8978b.png)

If you reduce it to only species with 2 or fewer radicals, you have  fewer of the top end.

Number of resonance structures :   Number of species of this type
                          1   :   457
                          2   :    82
                          3   :   481
                          4   :    40
                          5   :     6
                          6   :   256
                          7   :     1
                          8   :     4
                          9   :    82
                         12   :    11
                         16   :     1

so a constraint of this type should help speed up model generation.

**Comment by mliu49** [comment]
I see, that makes a lot of sense. Each of these species has 3 localized regions with 3 resonance forms and 1 region with 2 resonance forms, giving 3x3x3x2=54 total resonance structures.

I wonder if there's a good way to cut down on these besides limiting the number of radicals.
